File: QTest/UserObject.m
//
//  UserObject.m
//  QTest
//
//  Created by Shah Newaz Hossain on 10/18/16.
//  Copyright © 2016 Shah Newaz Hossain. All rights reserved.
//

#import "UserObject.h"

static UserObject *Shared = nil;

@implementation UserObject

+ (UserObject*) sharedUser {
    if (Shared==nil) {
        Shared = [[UserObject alloc] init];
        //        NSLog(@"creating user.............");
    }
    return Shared;
}

- (instancetype)initWithData:(NSDictionary *)dictionary {
    if (self = [super init]) {
        if (dictionary.allKeys > 0) {
            _userId = dictionary[@"user_id"];
            _userName = dictionary[@"user_name"];
            _testTakenCount = dictionary[@"test_taken_count"];
            _firstName = dictionary[@"first_name"];
            _email = dictionary[@"email"];
            if (dictionary[@"contact_number"]) {
                _contactNumber = dictionary[@"contact_number"];
            }
            _isAdmin = [dictionary[@"is_admin"] boolValue];
        }
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)decoder {
    if(self = [super init]) {
        _userId = [decoder decodeObjectForKey:@"user_id"];
        _userName = [decoder decodeObjectForKey:@"user_name"];
        _testTakenCount = [decoder decodeObjectForKey:@"test_taken_count"];
        _firstName = [decoder decodeObjectForKey:@"first_name"];
        _email = [decoder decodeObjectForKey:@"email"];
        _contactNumber = [decoder decodeObjectForKey:@"contact_number"];
        _isAdmin = [[decoder decodeObjectForKey:@"is_admin"] boolValue];

    }
    return self;
}

- (void)encodeWithCoder:(NSCoder *)aCoder {
    [aCoder encodeObject:_userId forKey:@"user_id"];
    [aCoder encodeObject:_userName forKey:@"user_name"];
    [aCoder encodeObject:_testTakenCount forKey:@"test_taken_count"];
    [aCoder encodeObject:_firstName forKey:@"first_name"];
    [aCoder encodeObject:_email forKey:@"email"];
    [aCoder encodeObject:_contactNumber forKey:@"contact_number"];
    [aCoder encodeObject:@(_isAdmin) forKey:@"is_admin"];
}

- (void)saveUserToLocalWithInfo:(NSDictionary*)info {
    UserObject *user = [[UserObject alloc] initWithData:info];
    _userId = user.userId;
    _userName = user.userName;
    _firstName = user.firstName;
    _email = user.email;
    _contactNumber = user.contactNumber;
    _testTakenCount = user.testTakenCount;
    _isAdmin = user.isAdmin;
    
    NSData *encodedData = [NSKeyedArchiver archivedDataWithRootObject:self];
    [USERDEFAULTS setObject:encodedData forKey:kStringSavedUserKey];
    [USERDEFAULTS setBool:YES forKey:kStringLoginKey];
    [USERDEFAULTS synchronize];
}
- (void)retriveUserFromLocal {
    NSData * data = [USERDEFAULTS objectForKey:kStringSavedUserKey];
    UserObject *user = [NSKeyedUnarchiver unarchiveObjectWithData:data];
    _userId = user.userId;
    _userName = user.userName;
    _firstName = user.firstName;
    _email = user.email;
    _contactNumber = user.contactNumber;
    _testTakenCount = user.testTakenCount;
    _isAdmin = user.isAdmin;
}

- (void)createContainerWithCentre:(UIViewController*)centreVC {
//    _container = [MFSideMenuContainerViewController containerWithCenterViewController:centreVC leftMenuViewController:nil rightMenuViewController:MenuItemsViewController.new];
//    _container.panMode = MFSideMenuPanModeNone;
//    _container.navigationItem.hidesBackButton = YES;
}

+ (BOOL)isValidEmail:(NSString*)email {
    NSPredicate *emailValidator = [NSPredicate predicateWithFormat:@"SELF MATCHES[c] %@",
                                   @"(?:[a-z0-9!#$%\\&'*+/=?\\^_`{|}~-]+(?:\\.[a-z0-9!#$%\\&'*+/=?\\^_`{|}"
                                   @"~-]+)*|\"(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21\\x23-\\x5b\\x5d-\\"
                                   @"x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])*\")@(?:(?:[a-z0-9](?:[a-"
                                   @"z0-9-]*[a-z0-9])?\\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\\[(?:(?:25[0-5"
                                   @"]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-"
                                   @"9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\\x01-\\x08\\x0b\\x0c\\x0e-\\x1f\\x21"
                                   @"-\\x5a\\x53-\\x7f]|\\\\[\\x01-\\x09\\x0b\\x0c\\x0e-\\x7f])+)\\])"];
    
    BOOL success = [emailValidator evaluateWithObject:email];
    return success;
}
@end


@implementation Question

- (instancetype)initWithData:(NSDictionary *)dictionary {
    if (self = [super init]) {
        if (dictionary.allKeys > 0) {
            _questionId = dictionary[@"question_id"];
            _questionText = dictionary[@"question_text"];
        }
    }
    return self;
}


@end


@implementation Category

- (instancetype)initWithData:(NSDictionary *)dictionary {
    if (self = [super init]) {
        if (dictionary.allKeys > 0) {
            _categoryId = dictionary[@"category_id"];
            _categoryTitle = dictionary[@"category_title"];
            _totalQuestions = dictionary[@"total_questions"];
            _totalCorrectAnswers = dictionary[@"total_correct_answers"];
            switch (_categoryId.integerValue) {
                case 2:
                    _bgColor = [UIColor colorWithRed:0.91 green:0.71 blue:0.35 alpha:1.00];
                    break;
                case 3:
                    _bgColor = [UIColor colorWithRed:0.55 green:0.76 blue:0.39 alpha:1.00];
                    break;
                case 4:
                    _bgColor = [UIColor colorWithRed:0.13 green:0.57 blue:0.65 alpha:1.00];
                    break;
                case 5:
                    _bgColor = [UIColor colorWithRed:0.51 green:0.68 blue:0.80 alpha:1.00];
                    break;
                case 6:
                    _bgColor = [UIColor colorWithRed:0.39 green:0.30 blue:0.58 alpha:1.00];
                    break;
                default:
                    _bgColor = [UIColor colorWithRed:0.82 green:0.32 blue:0.27 alpha:1.00];
                    break;
            }
        }
    }
    return self;
}

@end

File: QTest/UserObject.swift
//
//  UserObject.swift
//  Dollars
//
//  Created by Shah Newaz Hossain on 8/27/16.
//  Copyright © 2016 ShahNewaz. All rights reserved.
//

import UIKit
import MFSideMenu

class UserObject: NSObject, NSCoding {
    
    var userId: String = ""
    var userName: String = ""
    var firstName: String = ""
    var testTakenCount: String = ""
    var email: String = ""
    var contactNumber: String = ""

    var container: MFSideMenuContainerViewController? = nil
    var willGoForTest: Bool = false

    static let sharedUser : UserObject = {
        let instance = UserObject()
        return instance
    }()
    
    func createContainer(centreVC: UIViewController) {
        container = MFSideMenuContainerViewController.container(withCenter: centreVC, leftMenuViewController: nil, rightMenuViewController: MenuItemsViewController())
        container?.panMode = MFSideMenuPanModeNone
        container?.navigationItem.hidesBackButton = true
    }
    
    convenience init(_ dictionary: NSDictionary) {
        self.init()
        guard dictionary.allKeys.count>0 else {
            return
        }
        userId = dictionary["user_id"] as! String
        userName = dictionary["user_name"] as! String
        testTakenCount = dictionary["test_taken_count"] as! String
        firstName = dictionary["first_name"] as! String
        email = dictionary["email"] as! String
        if let contact_number = dictionary["contact_number"] as? String {
            contactNumber = contact_number
        }
        
    }
    
    
    func encode(with aCoder: NSCoder) {
        aCoder.encode(userId, forKey: "user_id")
        aCoder.encode(userName, forKey: "user_name")
        aCoder.encode(testTakenCount, forKey: "test_taken_count")
        aCoder.encode(firstName, forKey: "first_name")
        aCoder.encode(email, forKey: "email")
        aCoder.encode(contactNumber, forKey: "contact_number")
    }
    
    
    
    required convenience init(coder decoder: NSCoder) {
        self.init()
        userId = decoder.decodeObject(forKey: "user_id") as! String
        userName = decoder.decodeObject(forKey: "user_name") as! String
        testTakenCount = decoder.decodeObject(forKey: "test_taken_count") as! String
        firstName = decoder.decodeObject(forKey: "first_name") as! String
        email = decoder.decodeObject(forKey: "email") as! String
        contactNumber = decoder.decodeObject(forKey: "contact_number") as! String
    }
    
    
    func saveUserToLocal(info: NSDictionary) {
        let user = UserObject.init(info)
        userId = user.userId
        userName = user.userName
        firstName = user.firstName
        email = user.email
        contactNumber = user.contactNumber
        testTakenCount = user.testTakenCount

        let encodedData = NSKeyedArchiver.archivedData(withRootObject: self)
        USERDEFAULTS.set(encodedData, forKey: kStringSavedUserKey)
        USERDEFAULTS.set(true, forKey: kStringLoginKey)
        USERDEFAULTS.synchronize()
        
    }
    
    func retriveUserFromLocal() {
        let data = USERDEFAULTS.value(forKey: kStringSavedUserKey) as! Data
        let user = NSKeyedUnarchiver.unarchiveObject(with: data) as! UserObject
        userId = user.userId
        userName = user.userName
        firstName = user.firstName
        email = user.email
        contactNumber = user.contactNumber
        testTakenCount = user.testTakenCount

    }

    
    func isValidEmail(testStr:String) -> Bool {
        print("validate emilId: \(testStr)")
        let emailRegEx = "^(?:(?:(?:(?: )*(?:(?:(?:\\t| )*\\r\\n)?(?:\\t| )+))+(?: )*)|(?: )+)?(?:(?:(?:[-A-Za-z0-9!#$%&’*+/=?^_'{|}~]+(?:\\.[-A-Za-z0-9!#$%&’*+/=?^_'{|}~]+)*)|(?:\"(?:(?:(?:(?: )*(?:(?:[!#-Z^-~]|\\[|\\])|(?:\\\\(?:\\t|[ -~]))))+(?: )*)|(?: )+)\"))(?:@)(?:(?:(?:[A-Za-z0-9](?:[-A-Za-z0-9]{0,61}[A-Za-z0-9])?)(?:\\.[A-Za-z0-9](?:[-A-Za-z0-9]{0,61}[A-Za-z0-9])?)*)|(?:\\[(?:(?:(?:(?:(?:[0-9]|(?:[1-9][0-9])|(?:1[0-9][0-9])|(?:2[0-4][0-9])|(?:25[0-5]))\\.){3}(?:[0-9]|(?:[1-9][0-9])|(?:1[0-9][0-9])|(?:2[0-4][0-9])|(?:25[0-5]))))|(?:(?:(?: )*[!-Z^-~])*(?: )*)|(?:[Vv][0-9A-Fa-f]+\\.[-A-Za-z0-9._~!$&'()*+,;=:]+))\\])))(?:(?:(?:(?: )*(?:(?:(?:\\t| )*\\r\\n)?(?:\\t| )+))+(?: )*)|(?: )+)?$"
        let emailTest = NSPredicate(format:"SELF MATCHES %@", emailRegEx)
        let result = emailTest.evaluate(with: testStr)
        return result
    }
    
//    func isValidPhoneNumber(value: String) -> Bool {
//        let PHONE_REGEX = "^\\d{3}-\\d{3}-\\d{4}$"
//        let phoneTest = NSPredicate(format: "SELF MATCHES %@", PHONE_REGEX)
//        let result =  phoneTest.evaluate(with: value)
//        return result
//    }
}

class Question: NSObject {

    var questionId: String = ""
    var answer: Int?
    var questionText: String = ""

    convenience init(_ dictionary: NSDictionary) {
        self.init()
        questionId = dictionary["question_id"] as! String
        questionText = dictionary["question_text"] as! String
    }

}

class Category: NSObject {
    
    var categoryId: String = ""
    var categoryTitle: String = ""
    var totalCorrectAnswers: NSNumber = 0
    var totalQuestions: NSNumber = 0
    var bgColor: UIColor = UIColor.lightGray

    convenience init(_ dictionary: NSDictionary) {
        self.init()
        categoryId = dictionary["category_id"] as! String
        categoryTitle = dictionary["category_title"] as! String
        totalCorrectAnswers = dictionary["total_correct_answers"] as! NSNumber
        totalQuestions = dictionary["total_questions"] as! NSNumber
        
//        let cid = NumberFormatter().number(from: categoryId)
        switch categoryId {
        case "2":
            bgColor = UIColor(red:0.10, green:0.64, blue:0.11, alpha:1.00)
            break
        case "3":
            bgColor = UIColor(red:0.99, green:0.96, blue:0.21, alpha:1.00)
            break
        case "4":
            bgColor = UIColor(red:0.33, green:0.16, blue:0.45, alpha:1.00)
            break
        case "5":
            bgColor = UIColor(red:0.86, green:0.38, blue:0.13, alpha:1.00)
            break
        case "6":
            bgColor = UIColor(red:0.20, green:0.44, blue:0.71, alpha:1.00)
            break
        default:
            bgColor = UIColor(red:0.99, green:0.05, blue:0.11, alpha:1.00)
            break
        }
    }
    
//    func encode(with aCoder: NSCoder) {
//        aCoder.encode(categoryId, forKey: "user_id")
//        aCoder.encode(userName, forKey: "user_name")
//        aCoder.encode(categoryTitle, forKey: "test_taken_count")
//        aCoder.encode(totalCorrectAnswers, forKey: "first_name")
//        aCoder.encode(totalQuestions, forKey: "email")
//        aCoder.encode(contactNumber, forKey: "contact_number")
//    }
//    
//    
//    
//    required convenience init(coder decoder: NSCoder) {
//        self.init()
//        userId = decoder.decodeObject(forKey: "user_id") as! String
//        userName = decoder.decodeObject(forKey: "user_name") as! String
//        testTakenCount = decoder.decodeObject(forKey: "test_taken_count") as! String
//        firstName = decoder.decodeObject(forKey: "first_name") as! String
//        email = decoder.decodeObject(forKey: "email") as! String
//        contactNumber = decoder.decodeObject(forKey: "contact_number") as! String
//    }
}

